src/runtime_src/core/tools/xbmgmt2/OO_NIFD.cpp

```diff
@@ -19,24 +19,116 @@
+  } catch (const std::exception& e) {
+      xrt_core::send_exception_message(e.what(), "XBMGMT");
+  }
+
+  // const int NIFD_CHECK_STATUS = 8;
+  unsigned int statusReg = 0;
+  int result = 0;//dev->ioctl(fd.get(), NIFD_CHECK_STATUS, &statusReg);
+  if (result != 0)
+    throw xrt_core::error("Could not read status register");
+  else
+    std::cout << boost::format("Current NIFD status: 0x%x\n") % statusReg;
+}
+
+static void
+readback(const std::string& device, const std::string file)
+{
+  std::ifstream fin(file.c_str()) ;
+  if (!fin)
+    throw xrt_core::error(boost::str(boost::format("Could not open %s for reading") % file));
+
+  std::vector<unsigned int> hardwareFramesAndOffsets;
+  while (!fin.eof())
+  {
+    unsigned int nextNum = 0 ;
+    fin >> std::ws >> nextNum >> std::ws ;
+    hardwareFramesAndOffsets.push_back(nextNum);
+  }
+  fin.close();
+
+  auto dev = xrt_core::get_mgmtpf_device(xrt_core::utils::bdf2index(device));
+  xrt_core::scope_value_guard<int, std::function<void()>> fd { 0, nullptr };
+  try {
+      fd = dev->file_open("nifd_pri", O_RDWR); 
+  } catch (const std::exception& e) {
+      xrt_core::send_exception_message(e.what(), "XBMGMT");
+  }
+  
+  unsigned int numBits = static_cast<unsigned int>(hardwareFramesAndOffsets.size() / 2);
+  unsigned int resultWords = numBits % 32 ? numBits/32 + 1 : numBits/32 ;
+  std::vector<uint64_t> packet(1 + numBits * 2 + resultWords);
+  packet.push_back(numBits) ;
+  for (unsigned int i = 0 ; i < hardwareFramesAndOffsets.size() ; ++i)
+    packet.push_back(hardwareFramesAndOffsets[i]);
+
+  // const int NIFD_READBACK_VARIABLE = 3 ;
+  // const int NIFD_SWITCH_ICAP_TO_NIFD = 4 ;
+  // const int NIFD_SWITCH_ICAP_TO_PR = 5 ;
+  int result = 0 ;
+  // result = dev->ioctl(fd.get(), NIFD_SWITCH_ICAP_TO_NIFD);
+  if (result != 0)
+    throw xrt_core::error("Could not switch ICAP to NIFD control");
+
+  // result = dev->ioctl(fd.get(), NIFD_READBACK_VARIABLE, packet.data()) ;
+  // result |= dev->ioctl(fd.get(), NIFD_SWITCH_ICAP_TO_PR);
+  if (result != 0)
+    throw xrt_core::error("Could not readback variable!");
+  
+  std::cout << "Value read: " ;
+  for (unsigned int i = 0 ; i < resultWords ; ++i)
+    std::cout << boost::format("0x%x ") % (packet[1 + numBits*2 + i]);
+}
+
+}
+//end anonymous namespace
+
 
 // ----- C L A S S   M E T H O D S -------------------------------------------
 
 OO_NIFD::OO_NIFD( const std::string &_longName)
     : OptionOptions(_longName, "<add description>")
     , m_device("")
     , m_help(false)
+    , status(false)
+    , readback_file("")
 {
   setExtendedHelp("<add description>");
 
   m_optionsDescription.add_options()
     ("device,d", boost::program_options::value<decltype(m_device)>(&m_device), "The Bus:Device.Function (e.g., 0000:d8:00.0) device of interest")
-    ("status", "<add description>")
-    ("read-back", "<add description>")
+    ("status", boost::program_options::bool_switch(&status), "<add description>")
```

Comment:
Please add a description, especially since you are adding this functionality.

Reply:
Have to set up  a meeting with Jason to talk in detail about this :)